### DISCUSSION TITLE
Client-side memory leaks in 1.12

### DISCUSSION THREAD

**Discussion by jchung01** [ACGaming/UniversalTweaks#324] [Bugfix Suggestions]
EDIT: I'll update this list as I find more.

I've been doing some memory profiling on my client on a large modpack, and I figured I would list some mods in 1.12 that seem to be causing memory leaks by holding onto `WorldClient` and/or `EntityPlayerSP`, noticeable when changing dimensions or reconnecting (as that creates a new client/player):

- [x] **CB Multipart**: known leak, documented [here](https://github.com/TheCBProject/CBMultipart/pull/79).
- [x] **MrTJPCore**: pretty much the exact same leak as CB Multipart, documented [here](https://github.com/MrTJP/MrTJPCore/issues/32)
- [x] **New Tardis Mod**: leak in [ClientProxy](https://github.com/Spectre0987/TardisMod/blob/master/src/main/java/net/tardis/mod/proxy/ClientProxy.java#L164) due to adding players to a map that is unused (because this mod was in alpha for 1.12)
- [ ] Cosmetic Armor Reworked: possible leak in [PlayerRenderHandler](https://github.com/zlainsama/CosmeticArmorReworked/blob/1.12.2/src/main/java/lain/mods/cos/client/PlayerRenderHandler.java#L40). It uses a timed eviction cache which could be the issue (more recent version use weak keys)
- [ ] JourneyMap (ARR): Possible leak in `journeymap.client.task.multi.RenderSpec,` I don't have much to back this up though besides what JProfiler says
- [x] Bibliocraft: Not really a memory leak in the strict sense, but does hold the first WorldClient (for whatever reason) of the session on world load/connect due to `VersionCheck`

EDIT:
- [x] Simply Jetpacks: Leak in `SyncHandler` maps client side - Server side, these maps are cleaned up properly [here](https://github.com/Tomson124/SimplyJetpacks2/blob/1.12/src/main/java/tonius/simplyjetpacks/handler/SyncHandler.java#L118) and [here](https://github.com/Tomson124/SimplyJetpacks2/blob/1.12/src/main/java/tonius/simplyjetpacks/handler/SyncHandler.java#L123), but client side also [updates these maps](https://github.com/Tomson124/SimplyJetpacks2/blob/1.12/src/main/java/tonius/simplyjetpacks/client/handler/KeybindHandler.java#L87) and those events are never called client-side, so the client-side maps are never cleaned up.
- [x] OpenComputers: Network leak for all(?) packets due to closing the `ByteBufInputStream` of the data, but not releasing the actual data.
- [x] CodeChicken Lib: Network leak for `PacketCustom` due to not releasing its data.
- [x] Iron Chests: Retains first `WorldClient` each on first time rendering crystal chest or crystal shulker box due to using a reusable `EntityItem`
- [x] Compact Machines: Retains client's current world when each miniaturization recipe is first viewed
- [x] Ender IO: Retains first `WorldClient` for each obelisk renderer due to using a reusable `EntityItem`

The first 3 (bolded) were the most consistent in holding the WorldClient, and although Tardis should be easy enough to fix (by just disabling adding to the map), I have tried a few ways to fix CB Multipart and MrTJPCore, but they are *in Scala*, so even if a simple fix is to use a WeakHashMap, I don't know how to apply it externally.
Some screenshots from a heap dump:
![heapdump](https://github.com/ACGaming/UniversalTweaks/assets/60687097/c4d1539f-07c2-4156-8598-ace6310ced04)
![heapdump2](https://github.com/ACGaming/UniversalTweaks/assets/60687097/4aa01538-00e5-49d4-969c-66a9b5ad7d28)

**Comment by jchung01** [comment]
Other memory leaks that recently got fixes by RLMixins that can be added to UT as well as I'm still not sure if it's compatible with MixinBooter 8:
- BetterNether: see [this](https://github.com/fonnymunkey/RLMixins/commit/de3d3212e1b5114e5a186d7cc6aa9c53e30f50d0#diff-83fdb383314652a2ad72eab2fd87f5bec1b5e3a32cc874d0fc17ca67915e2417)
- BetterQuesting: see [this](https://github.com/fonnymunkey/RLMixins/commit/de3d3212e1b5114e5a186d7cc6aa9c53e30f50d0#diff-a4a9071944a8fd07ecf8c2350d07f61a8b0f7e7fa4295817c9b93c5823d73b03)
- MoBends: various fixes, see [MoBends files](https://github.com/fonnymunkey/RLMixins/commit/de3d3212e1b5114e5a186d7cc6aa9c53e30f50d0#diff-d74c3185c4958d2f1f29d6245ed437bdb881fceb0855be485bcd893614b04b3b)

**Comment by ACGaming** [comment]
Highly doubt that mixins for Scala imports would be easy, so I included them in separate forks.

**Comment by jchung01** [comment]
I think there *may* be a way to properly cleanup the maps using packets, so I'll look into it while looking into SimplyJetpacks.

**Comment by ACGaming** [comment]
Other approaches might be possible, yes. But don‘t want to mess with Scala more than needed!

**Comment by jchung01** [comment]
Added fix for TARDIS and Simply Jetpacks in above PR. It would indeed require messing with a lot more files in Scala with MrTJPCore and CBMultipart and their dependents, so looks like the forks are the way to go.

**Comment by ACGaming** [comment]
Maybe we could add this to the list as well?
https://github.com/xJon/Tekkit-2/issues/43

**Reply by jchung01** [reply]
@xJon Should be addressed in #496. Hopefully there's no bugs with my fix.

**Reply by xJon** [reply]
@jchung01 Thank you 💯 (for some reason GitHub did not send a notification)
